scipy/_lib/_array_api_util.py

```diff
@@ -0,0 +1,64 @@
+"""Utility functions to use Python Array API compatible libraries.
+
+For the context about the Array API see:
+https://data-apis.org/array-api/latest/purpose_and_scope.html
+
+The SciPy use case of the Array API is described on the following page:
+https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy
+"""
+
+import numpy as np
+
+from scipy._lib import _pep440
+
+_NUMPY_INCLUDES_ARRAY_API = _pep440.parse(np.__version__) >= _pep440.Version(
+    "1.22.0"
+)
+
+
+def _get_namespace(*xs):
+    """
+    Returns the module that implements Array API compatible functions or NumPy.
+
+    Parameters
+    ----------
+    *xs : sequence of array_like
+        Arrays to get the namespace from.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x = np.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy'
+
+    >>> import numpy.array_api as xp
+    >>> x = xp.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy.array_api'
+    """
+    # `xs` contains one or more arrays
+    # When there is no __array_namespace__, we will use assume the default (NumPy) namespace.
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, "__array_namespace__") else np
+        for x in xs
+    }
+
+    if len(namespaces) != 1:
+        raise ValueError(
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
+        )
+
+    (xp,) = namespaces
+
+    return xp
+
+
+def _concatenate(arrays, axis):
```

Comment:
I like this pattern. Creating a wrapper and using it as a drop in replacement across the library. It should be quite easy to deprecate in the distant future when `numpy.array_api` will stabilise.

Reply:
Yes, but maybe it would be good to follow a common naming scheme to facilitate this and be clearer about where this is coming from.

Reply:
I suppose we should try to follow a similar path to `scikit-learn`. There's a long discussion here that Pamphile linked from the issue: https://github.com/scikit-learn/scikit-learn/issues/22352

Are we converging to similar approaches now?

Reply:
I also saw `scikit-learn` attempt to use NEP37 like https://github.com/scikit-learn/scikit-learn/pull/16574. It is hard to keep track of all the approaches. There's also NEP47 approach used here of course: https://numpy.org/neps/nep-0047-array-api-standard.html#appendix-a-possible-get-namespace-implementation

So, it is pretty much decided to drop/not accept NEP37 approach then? I initially went down that path because I had noticed that CuPy already had https://docs.cupy.dev/en/stable/reference/generated/cupy.get_array_module.html.